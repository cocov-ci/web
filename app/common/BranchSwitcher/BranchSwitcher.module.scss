@use "sass:color";
@use "sass:map";
@import "styles/variables";

.base {
  box-shadow: 7px 2px 40px 0 color.adjust($colorBlack, $alpha: -0.5);
  border-radius: 8px;
  width: 290px;
  height: 310px;
  position: absolute;
  backdrop-filter: blur(10px) saturate(100%);

  .background {
    position: absolute;
    width: 100%;
    height: 100%;
    background: color.adjust($colorBlack, $alpha: -0.24);
    border-radius: 8px;
  }

  .content {
    position: relative;

    .header {
      display: flex;
      justify-content: space-between;
      padding: 12px;

      .title {
        color: $colorGray;
        font-size: 17px;
        font-variation-settings: "wght" 600;
      }

      .close {
        display: flex;

        .closeButton {
          cursor: pointer;
          margin-left: 8px;
          border: 0;
          background: transparent;
          color: $colorGray;
          padding: 0;
          height: 24px;
          width: 24px;
          display: block;
        }
      }
    }

    .searchContainer {
      margin-top: 2px;
      padding: 0 12px;

      .searchInput {
        height: 32px;

        input {
          font-size: 15px;
        }
      }
    }

    .itemsContainer {
      display: flex;
      flex: 1;
      overflow: scroll;
      flex-direction: column;
      padding: 0 12px;
      margin-top: 12px;

      .loading {
        height: 35px;
        display: flex;
        margin-bottom: 3px;

        span {
          width: 100%;
        }
      }

      .item {
        padding: 6px 12px;
        color: $colorGray;
        margin-bottom: 3px;
        border: 1px solid transparent;

        &.active {
          background: map.get($colorBranchSwitcher, "active-item-background");
          border: 1px solid map.get($colorBranchSwitcher, "active-item-border");
          border-radius: 8px;
        }

        &:hover {
          background: color.adjust(map.get($colorBranchSwitcher, "active-item-background"), $alpha: -0.5);
          border: 1px solid color.adjust(map.get($colorBranchSwitcher, "active-item-border"), $alpha: -0.5);
          border-radius: 8px;
        }
      }
    }
  }
}
