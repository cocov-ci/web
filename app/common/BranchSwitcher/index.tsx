'use client'

import classNames from 'classnames'
import { Search, X } from 'lucide-react'
import { useRouter } from 'next/navigation'
import React, { useEffect, useMemo, useRef, useState } from 'react'

import Input from 'app/common/Input'
import Kbd from 'app/common/Kbd'
import Loading from 'app/common/Loading'
import { useSegments } from 'context/SegmentsContext'
import useFetch from 'hooks/useFetch'
import useOnClickOutside from 'hooks/useOnClickOutside'
import randomBetween from 'utils/randomBetween'

import styles from './BranchSwitcher.module.scss'

type BranchSwitcherProps = {
  className?: string
  onClose: () => void
}

interface BranchesFetchResponse {
  data: string[]
  loading: boolean
}

const BranchSwitcher = ({ className, onClose }: BranchSwitcherProps) => {
  const ref = useRef(null)
  const router = useRouter()
  const segments = useSegments()
  const [selectedItem, setSelectedItem] = useState<number>(0)
  const [results, setResults] = useState<Array<string> | undefined>()
  const repositoryName = useMemo(() => segments[1], [segments])

  useOnClickOutside(ref, onClose)

  const { data: branches, loading } = useFetch({
    url: `/api/repositories/${repositoryName}/branches`,
    handler: [],
  }) as BranchesFetchResponse

  const onSearchChange = (term: string) => {
    setResults(branches.filter(item => item.includes(term)))
  }

  const onSelectBranch = (selectedItem: string) => {
    router.push(`repos/${repositoryName}/branches/${selectedItem}`)
    onClose && onClose()
  }

  const handleKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (!results || results.length === 0) {
      return
    }

    switch (event.key) {
      case 'ArrowUp':
        if (selectedItem === 0) {
          setSelectedItem(results.length - 1)
        } else {
          setSelectedItem(selectedItem - 1)
        }

        break

      case 'ArrowDown':
        if (selectedItem === results.length - 1) {
          setSelectedItem(0)
        } else {
          setSelectedItem(selectedItem + 1)
        }

        break

      case 'Escape':
        onClose && onClose()

        break

      case 'Enter':
        onSelectBranch(results[selectedItem])

        break
    }
  }

  useEffect(() => {
    setResults(branches)
  }, [branches])

  return (
    <div className={classNames(styles.base, className)} ref={ref}>
      <div className={styles.background} />

      <div className={styles.content}>
        <div className={styles.header}>
          <span className={styles.title}>Switch Branch</span>
          <div className={styles.close}>
            <Kbd size="mini" text="esc" variation="dark" />
            <button
              className={styles.closeButton}
              onClick={() => onClose && onClose()}
            >
              <X size={24} />
            </button>
          </div>
        </div>
        <div className={styles.searchContainer}>
          <Input
            autoFocus={true}
            className={styles.searchInput}
            icon={Search}
            onChange={event =>
              branches.length > 0 && onSearchChange(event.target.value)
            }
            onKeyUp={handleKeyUp}
            placeholder="Type to search..."
            type="text"
            variation="dark"
          />
        </div>
        <div className={styles.itemsContainer}>
          {loading &&
            new Array(randomBetween(1, 5)).fill(0).map((_, idx) => (
              <Loading
                className={styles.loading}
                height="35px"
                // eslint-disable-next-line react/no-array-index-key
                key={idx}
                type="skeleton"
                variation="dark"
                width="100%"
              />
            ))}

          {!loading &&
            results?.map((name, idx) => (
              <button
                className={classNames(styles.item, {
                  [styles.active]: selectedItem == idx,
                })}
                key={name}
                onClick={e => {
                  e.preventDefault()
                  onSelectBranch(name)
                }}
              >
                {name}
              </button>
            ))}
        </div>
      </div>
    </div>
  )
}

export default BranchSwitcher
