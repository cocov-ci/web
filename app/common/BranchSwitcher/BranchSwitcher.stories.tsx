import type { Meta, StoryObj } from '@storybook/react'
import React from 'react'

import BackgroundCats from 'stories/BackgroundCats'
import { satoshi } from 'utils/fonts'

import BranchSwitcher from './index'

const meta: Meta<typeof BranchSwitcher> = {
  title: 'Common/BranchSwitcher',
  component: BranchSwitcher,
  tags: ['autodocs'],
  decorators: [
    Story => (
      <div
        className={satoshi.className}
        style={{ position: 'relative', margin: '0', padding: '0' }}
      >
        <BackgroundCats />
        <div style={{ margin: '0', padding: '100px' }}>
          <Story />
        </div>
      </div>
    ),
  ],
}

export default meta

type Story = StoryObj<typeof BranchSwitcher>

export const Default: Story = {
  args: {
    visible: true,
  },
}
