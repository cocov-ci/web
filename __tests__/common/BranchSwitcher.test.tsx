import { fireEvent, render, screen } from '@testing-library/react'

import BranchSwitcher from 'app/common/BranchSwitcher'
import '@testing-library/jest-dom'

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}))

describe('common/BranchSwitcher', () => {
  const { container } = render(
    <BranchSwitcher onClose={() => null} visible={true} />,
  )

  it('renders BranchSwitcher', () => {
    expect(screen.getByRole('textbox')).toBeVisible()
  })

  it('triggers onClose event', () => {
    const handleClick = jest.fn()

    render(<BranchSwitcher onClose={handleClick} visible={true} />)
    fireEvent.click(screen.getByRole('button'))
    expect(handleClick).toHaveBeenCalledTimes(1)
  })

  it('triggers onKeyUp event', () => {
    const handleOnClose = jest.fn()
    render(<BranchSwitcher onClose={handleOnClose} visible={true} />)

    const input = screen.getByRole('textbox') as HTMLInputElement
    fireEvent.keyUp(input, { key: 'Escape', code: 'Escape' })

    expect(handleOnClose).toHaveBeenCalledTimes(1)
  })

  it('renders BranchSwitcher snapshots', () => {
    expect(container).toMatchSnapshot()
  })
})
