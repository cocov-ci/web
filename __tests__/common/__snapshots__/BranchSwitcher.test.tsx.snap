// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/BranchSwitcher renders BranchSwitcher snapshots 1`] = `<div />`;
